db/blob/blob_source.cc

```diff
@@ -0,0 +1,295 @@
+      lowest_used_cache_tier_(immutable_options->lowest_used_cache_tier) {}
+
+BlobSource::~BlobSource() { delete blob_file_cache_; }
+
+Status BlobSource::MaybeReadBlobFromCache(
+    FilePrefetchBuffer* prefetch_buffer, const ReadOptions& read_options,
+    const CacheKey& cache_key, uint64_t offset, const bool wait,
+    CachableEntry<Slice>* blob_entry) const {
+  assert(blob_entry != nullptr);
+  assert(blob_entry->IsEmpty());
+
+  // First, try to get the blob from the cache
+  //
+  // If either blob cache is enabled, we'll try to read from it.
+  Status s;
+  Slice key;
+  if (blob_cache_ != nullptr) {
+    key = cache_key.AsSlice();
+    s = GetBlobFromCache(key, blob_cache_.get(), blob_entry, wait);
+    if (blob_entry->GetValue()) {
+      if (prefetch_buffer) {
+        // Update the blob details so that PrefetchBuffer can use the read
+        // pattern to determine if reads are sequential or not for prefetching.
+        // It should also take in account blob read from cache.
+        prefetch_buffer->UpdateReadPattern(
+            offset, blob_entry->GetValue()->size(),
+            read_options.adaptive_readahead /* decrease_readahead_size */);
+      }
+    }
+  }
+
+  assert(s.ok() || blob_entry->GetValue() == nullptr);
+
+  return s;
+}
+
+Cache::Handle* BlobSource::GetEntryFromCache(
+    const CacheTier& cache_tier, Cache* blob_cache, const Slice& key,
+    const bool wait, const Cache::CacheItemHelper* cache_helper,
+    const Cache::CreateCallback& create_cb, Cache::Priority priority) const {
+  assert(blob_cache);
+
+  Cache::Handle* cache_handle = nullptr;
+
+  if (cache_tier == CacheTier::kNonVolatileBlockTier) {
+    cache_handle = blob_cache->Lookup(key, cache_helper, create_cb, priority,
+                                      wait, statistics_);
+  } else {
+    cache_handle = blob_cache->Lookup(key, statistics_);
+  }
+
+  return cache_handle;
+}
+
+Status BlobSource::InsertEntryToCache(
+    const CacheTier& cache_tier, Cache* blob_cache, const Slice& key,
+    const Cache::CacheItemHelper* cache_helper, const Slice* blob,
+    size_t charge, Cache::Handle** cache_handle,
+    Cache::Priority priority) const {
+  Status s;
+
+  if (cache_tier == CacheTier::kNonVolatileBlockTier) {
+    s = blob_cache->Insert(
+        key, const_cast<void*>(static_cast<const void*>(blob->data())),
+        cache_helper, charge, cache_handle, priority);
+  } else {
+    s = blob_cache->Insert(
+        key, const_cast<void*>(static_cast<const void*>(blob->data())), charge,
+        cache_helper->del_cb, cache_handle, priority);
+  }
+
+  return s;
+}
+
+Status BlobSource::GetBlobFromCache(const Slice& cache_key, Cache* blob_cache,
+                                    CachableEntry<Slice>* blob,
+                                    bool wait) const {
+  assert(blob);
+  assert(blob->IsEmpty());
+  const Cache::Priority priority = Cache::Priority::LOW;
+
+  // Lookup uncompressed cache first
```

Comment:
Minor but some of these comments seem to be left over/are not relevant; we won't have to add a separate "compressed blob cache" since that role can be filled by the secondary cache

Reply:
okay, done.